pkg/autoscaler/aggregation/bucketing.go

```diff
@@ -24,43 +24,51 @@ import (
-	// granularity has expired since the last write) so that we can zero those
-	// entries in the buckets array. It is also used when calculating the
-	// WindowAverage to know how much of the buckets array represents valid data.
-	lastWrite time.Time
-
-	// granularity is the duration represented by each bucket in the buckets ring buffer.
-	granularity time.Duration
-	// window is the total time represented by the buckets ring buffer.
-	window time.Duration
-	// The total sum of all buckets within the window. This total includes
-	// invalid buckets, e.g. buckets written to before firstTime or after
-	// lastTime are included in this total.
-	windowTotal float64
-
-	// decayMultiplier contains the speed with which the importance
-	// of items in the past decays. The larger the faster weights decay.
-	// It is autocomputed from window size and weightPrecision constant
-	// and is bounded by minExponent below.
-	decayMultiplier float64
-}
+type (
+	// TimedFloat64Buckets keeps buckets that have been collected at a certain time.
+	TimedFloat64Buckets struct {
+		bucketsMutex sync.RWMutex
+		// buckets is a ring buffer indexed by timeToIndex() % len(buckets).
+		// Each element represents a certain granularity of time, and the total
+		// represented duration adds up to a window length of time.
+		buckets []float64
+
+		// firstWrite holds the time when the first write has been made.
+		// This time is reset to `now` when the very first write happens,
+		// or when a first write happens after `window` time of inactivity.
+		// The difference between `now` and `firstWrite` is used to compute
+		// the number of eligible buckets for computation of average values.
+		firstWrite time.Time
+
+		// lastWrite stores the time when the last write was made.
+		// This is used to detect when we have gaps in the data (i.e. more than a
+		// granularity has expired since the last write) so that we can zero those
+		// entries in the buckets array. It is also used when calculating the
+		// WindowAverage to know how much of the buckets array represents valid data.
+		lastWrite time.Time
+
+		// granularity is the duration represented by each bucket in the buckets ring buffer.
+		granularity time.Duration
+		// window is the total time represented by the buckets ring buffer.
+		window time.Duration
+		// The total sum of all buckets within the window. This total includes
+		// invalid buckets, e.g. buckets written to before firstTime or after
+		// lastTime are included in this total.
+		windowTotal float64
+	}
+
+	// WeightedFloat64Buckets is the implementation of buckets, that
+	// uses weighted average algorithm.
+	WeightedFloat64Buckets struct {
+		*TimedFloat64Buckets
+
+		// decayMultiplier contains the speed with which the importance
+		// of items in the past decays. The larger the faster weights decay.
+		// It is autocomputed from window size and weightPrecision constant
+		// and is bounded by minExponent below.
+		decayMultiplier float64
```

Comment:
being SUPER pedantic: Im not totally sure `multiplier` captures what this does. But.. Im also not sure I have a better suggestion 😅. Maybe it's fine since it's private, but if there's anything better you can think of that'd be nice I think (I'm struggling to get an intuition what this field does other than reading the test and thinking "that seems like it does something reasonable", and it'd be nice if I could).

Reply:
I'll think about it :)

Reply:
went with smoothing/